File: app/src/main/java/com/attilapalfi/exceptional/persistence/ExceptionTypeStore.java
package com.attilapalfi.exceptional.persistence;

import java.util.*;

import android.os.Handler;
import android.os.Looper;
import com.attilapalfi.exceptional.dependency_injection.Injector;
import com.attilapalfi.exceptional.interfaces.VotedTypeListener;
import com.attilapalfi.exceptional.model.ExceptionType;
import io.paperdb.Book;
import io.paperdb.Paper;

import static java8.util.stream.StreamSupport.stream;

/**
 * Created by Attila on 2015-06-09.
 */
public class ExceptionTypeStore {
    private static final String TYPE_DATABASE = "TYPE_DATABASE";
    private static final ExceptionType EMPTY_TYPE = new ExceptionType( 0, "", "", "" );
    private final String MAX_ID = "maxId";
    private final String HAS_DATA = "hasData";
    private volatile boolean hasData = false;

    private Book database;
    private Handler handler;
    private final Map<String, List<ExceptionType>> exceptionTypeStore = Collections.synchronizedMap( new HashMap<>() );
    private final List<ExceptionType> votedExceptionTypeList = Collections.synchronizedList( new LinkedList<>() );
    private Set<VotedTypeListener> votedTypeListeners = new HashSet<>();

    public ExceptionTypeStore( ) {
        Injector.INSTANCE.getApplicationComponent().inject( this );
        database = Paper.book( TYPE_DATABASE );
        handler = new Handler( Looper.getMainLooper() );
        synchronized ( exceptionTypeStore ) {
            if ( database.read( HAS_DATA, false ) ) {
                initExceptionTypeStore();
                sortExceptionStore();
                hasData = true;
            }
        }
    }

    public void addExceptionTypes( List<ExceptionType> exceptionTypes ) {
        synchronized ( exceptionTypeStore ) {
            saveExceptionTypeStore( exceptionTypes );
        }
        hasData = true;
        database.write( HAS_DATA, true );
    }

    private void saveExceptionTypeStore( List<ExceptionType> exceptionTypes ) {
        int maxId = 0;
        for ( ExceptionType exception : exceptionTypes ) {
            if ( !exceptionTypeStore.containsKey( exception.getType() ) ) {
                exceptionTypeStore.put( exception.getType(), Collections.synchronizedList( new LinkedList<>() ) );
            }
            exceptionTypeStore.get( exception.getType() ).add( exception );
            database.write( Integer.toString( exception.getId() ), exception );
            maxId = exception.getId() > maxId ? exception.getId() : maxId;
        }
        database.write( MAX_ID, maxId );
        sortExceptionStore();
    }

    public void setVotedExceptionTypes( List<ExceptionType> exceptionTypes ) {
        synchronized ( votedExceptionTypeList ) {
            votedExceptionTypeList.clear();
            votedExceptionTypeList.addAll( exceptionTypes );
            sortVotedExceptionList();
        }
        handler.post( this::notifyVotedTypeListeners );
    }

    private void initExceptionTypeStore( ) {
        int minId = 1;
        int maxId = database.read( MAX_ID, 0 );
        for ( int i = minId; i <= maxId; i++ ) {
            ExceptionType exceptionType = database.read( Integer.toString( i ), EMPTY_TYPE );
            if ( !exceptionTypeStore.containsKey( exceptionType.getType() ) ) {
                exceptionTypeStore.put( exceptionType.getType(), Collections.synchronizedList( new LinkedList<>() ) );
            }
            exceptionTypeStore.get( exceptionType.getType() ).add( exceptionType );
        }
    }

    private void sortExceptionStore( ) {
        for ( List<ExceptionType> typeList : exceptionTypeStore.values() ) {
            Collections.sort( typeList, new ExceptionType.ShortNameComparator() );
        }
    }

    private void sortVotedExceptionList( ) {
        Collections.sort( votedExceptionTypeList, new ExceptionType.VoteComparator() );
    }

    public ExceptionType findById( int id ) {
        Map<String, List<ExceptionType>> tempMap = new HashMap<>( exceptionTypeStore );
        for ( List<ExceptionType> exceptionTypeList : new ArrayList<>( tempMap.values() ) ) {
            for ( ExceptionType exceptionType : exceptionTypeList ) {
                if ( id == exceptionType.getId() ) {
                    return exceptionType;
                }
            }
        }
        return new ExceptionType();
    }

    public Set<String> getExceptionTypes( ) {
        Map<String, List<ExceptionType>> tempMap = new HashMap<>( exceptionTypeStore );
        return new HashSet<>( tempMap.keySet() );
    }

    public List<ExceptionType> getExceptionTypeListByName( String typeName ) {
        Map<String, List<ExceptionType>> tempMap = new HashMap<>( exceptionTypeStore );
        if ( tempMap.containsKey( typeName ) ) {
            return new ArrayList<>( tempMap.get( typeName ) );
        }
        return new ArrayList<>();
    }

    public void wipe( ) {
        exceptionTypeStore.clear();
        votedExceptionTypeList.clear();
        database.destroy();
    }

    public List<ExceptionType> getVotedExceptionTypeList( ) {
        return new ArrayList<>( votedExceptionTypeList );
    }

    public void updateVotedType( ExceptionType votedType ) {
        synchronized ( votedExceptionTypeList ) {
            int index = votedExceptionTypeList.indexOf( votedType );
            ExceptionType listElementException = votedExceptionTypeList.get( index );
            listElementException.setVoteCount( votedType.getVoteCount() );
        }
        notifyVotedTypeListeners();
    }

    public void addVotedType( ExceptionType submittedType ) {
        synchronized ( votedExceptionTypeList ) {
            votedExceptionTypeList.add( submittedType );
        }
        notifyVotedTypeListeners();
    }

    public void notifyVotedTypeListeners( ) {
        if ( Looper.myLooper() == Looper.getMainLooper() ) {
            stream( votedTypeListeners ).forEach( VotedTypeListener::onVoteListChanged );
        }
    }

    public boolean addVotedTypeListener( VotedTypeListener listener ) {
        return votedTypeListeners.add( listener );
    }

    public boolean removeVotedTypeListener( VotedTypeListener listener ) {
        return votedTypeListeners.remove( listener );
    }

    public boolean hasData() { return hasData; }
}

File: app/src/main/java/com/attilapalfi/exceptional/persistence/ExceptionTypeStore.kt
package com.attilapalfi.exceptional.persistence

import android.os.Handler
import android.os.Looper
import com.attilapalfi.exceptional.dependency_injection.Injector
import com.attilapalfi.exceptional.interfaces.VotedTypeListener
import com.attilapalfi.exceptional.model.ExceptionType
import io.paperdb.Book
import io.paperdb.Paper
import java.util.*

/**
 * Created by Attila on 2015-06-09.
 */
public class ExceptionTypeStore : AbstractStore() {
    private val MAX_ID = "maxId"
    private val HAS_DATA = "hasData"
    @Volatile
    private var hasData = false
    @Volatile
    override public var initialized = false

    private val database: Book
    private val handler: Handler
    private val exceptionTypeStore = Collections.synchronizedMap(HashMap<String, LinkedList<ExceptionType>>())
    private val votedExceptionTypeList = Collections.synchronizedList(LinkedList<ExceptionType>())
    private val votedTypeListeners = HashSet<VotedTypeListener>()

    init {
        Injector.INSTANCE.applicationComponent.inject(this)
        database = Paper.book(TYPE_DATABASE)
        handler = Handler(Looper.getMainLooper())

    }

    override public fun init() {
        synchronized (exceptionTypeStore) {
            if (database.read(HAS_DATA, false)) {
                initExceptionTypeStore()
                sortExceptionStore()
                hasData = true
            }
            initialized = true
        }
    }

    public fun addExceptionTypes(exceptionTypes: List<ExceptionType>) {
        synchronized (exceptionTypeStore) {
            saveExceptionTypeStore(exceptionTypes)
        }
        hasData = true
        database.write(HAS_DATA, true)
    }

    private fun saveExceptionTypeStore(exceptionTypes: List<ExceptionType>) {
        var maxId = 0
        for (exception in exceptionTypes) {
            if (!exceptionTypeStore.containsKey(exception.type)) {
                exceptionTypeStore.put(exception.type, LinkedList<ExceptionType>())
            }
            exceptionTypeStore[exception.type]?.add(exception)
            database.write(Integer.toString(exception.id), exception)
            maxId = if (exception.id > maxId) exception.id else maxId
        }
        database.write(MAX_ID, maxId)
        sortExceptionStore()
    }

    public fun setVotedExceptionTypes(exceptionTypes: List<ExceptionType>) {
        synchronized (votedExceptionTypeList) {
            votedExceptionTypeList.clear()
            votedExceptionTypeList.addAll(exceptionTypes)
            sortVotedExceptionList()
        }
        handler.post { this.notifyVotedTypeListeners() }
    }

    private fun initExceptionTypeStore() {
        val minId = 1
        val maxId = database.read(MAX_ID, 0)
        for (i in minId..maxId) {
            val exceptionType = database.read(Integer.toString(i), EMPTY_TYPE)
            if (!exceptionTypeStore.containsKey(exceptionType.type)) {
                exceptionTypeStore.put(exceptionType.type, LinkedList<ExceptionType>())
            }
            exceptionTypeStore[exceptionType.type]?.add(exceptionType)
        }
    }

    private fun sortExceptionStore() {
        for (typeList in exceptionTypeStore.values) {
            Collections.sort(typeList, ExceptionType.ShortNameComparator())
        }
    }

    private fun sortVotedExceptionList() {
        Collections.sort(votedExceptionTypeList, ExceptionType.VoteComparator())
    }

    public fun findById(id: Int): ExceptionType {
        waitTillInitialized()
        synchronized(exceptionTypeStore) {
            val tempMap = HashMap(exceptionTypeStore)
            for (exceptionTypeList in ArrayList(tempMap.values)) {
                for (exceptionType in exceptionTypeList) {
                    if (id == exceptionType.id) {
                        return exceptionType
                    }
                }
            }
            return ExceptionType()
        }
    }

    public fun getExceptionTypes(): Set<String> {
        synchronized(exceptionTypeStore) {
            val tempMap = HashMap(exceptionTypeStore)
            return HashSet(tempMap.keys)
        }
    }

    public fun getExceptionTypeListByName(typeName: String): List<ExceptionType> {
        synchronized(exceptionTypeStore) {
            val tempMap = HashMap(exceptionTypeStore)
            if (tempMap.containsKey(typeName)) {
                return ArrayList(tempMap[typeName])
            }
            return ArrayList()
        }
    }

    public fun getVotedExceptionTypeList(): List<ExceptionType> {
        synchronized(exceptionTypeStore) {
            return ArrayList(votedExceptionTypeList)
        }
    }

    public fun updateVotedType(votedType: ExceptionType) {
        synchronized (votedExceptionTypeList) {
            val index = votedExceptionTypeList.indexOf(votedType)
            val listElementException = votedExceptionTypeList.get(index)
            listElementException.voteCount = votedType.voteCount
        }
        notifyVotedTypeListeners()
    }

    public fun addVotedType(submittedType: ExceptionType) {
        synchronized (votedExceptionTypeList) {
            votedExceptionTypeList.add(submittedType)
        }
        notifyVotedTypeListeners()
    }

    public fun notifyVotedTypeListeners() {
        if (Looper.myLooper() == Looper.getMainLooper()) {
            votedTypeListeners.forEach { it.onVoteListChanged() }
        }
    }

    public fun addVotedTypeListener(listener: VotedTypeListener): Boolean {
        return votedTypeListeners.add(listener)
    }

    public fun removeVotedTypeListener(listener: VotedTypeListener): Boolean {
        return votedTypeListeners.remove(listener)
    }

    public fun hasData(): Boolean {
        return hasData
    }

    public fun wipe() {
        exceptionTypeStore.clear()
        votedExceptionTypeList.clear()
        database.destroy()
    }

    companion object {
        private val TYPE_DATABASE = "TYPE_DATABASE"
        private val EMPTY_TYPE = ExceptionType(0, "", "", "")
    }
}
